contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,174 @@
+# Kubernetes Gang-Scheduling
+
+**Authors**: @k82cn , **Shepherds**: @vishh, @bsalamat
+
+**SIG/WG**: sig-scheduling, ML-WG
+
+## Motivation
+
+After the discussion at [Gang-scheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv) proposal, we decide to make Gang-scheduling API out of core by CRD, and implement it in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator). kube-arbitrator focus on "batch" workload in kubernetes, and will share the same [scheduling frameworks](https://github.com/kubernetes/community/pull/2281) when it's ready. This document is used to provide definition of CRD for gang-scheduling, and suggestion of customized job controllers.
+
+## Function Detail
+
+### CRD Definition
```

Comment:
I have a slightly different idea about the API. Can we create a **namespaced** object, let's call it a gang, that only specifies properties of a gang. Something like:

```go
type Gang struct {
    metav1.TypeMeta
    metav1.ObjectMeta

    NumMembers int
    Policy GangLifeCyclePolicy
}
```

Then Pod.PodSpec can have a field called `GangName` or `GangId`. When the field is not empty, it means that the pod is gang member. When scheduling pods which are gang members, scheduler makes sure that at least `Gang.NumMember` of such pods can be scheduled simultaneously or none of them are scheduled.

The Gang type can also include gang life cycle policies. For example, the policy can specify that when the number of running pods drops below `NumMember` kill all the other instances, or wait for a certain amount of time for the gang members to reach `NumMember` before killing other instances, ....

Here are some of the benefits of this design:
1. Existing workloads that use one of the K8s collections, e.g. ReplicaSet, Job, etc., can keep their config and only add the `GangName` to their pod template.
1. Flexibility. Pods of a gang can have different PodSpec and belong to different collections.
1. Existing controllers which are responsible for managing life cycle of collections work well with this design.
1. Users who have already learned and worked with K8s collections can keep using the familiar collections.

Reply:
This sounds like something that should be a label before we create a field on pod spec (also gang wouldn’t be a great name for an api object).

Reply:
We need a way to identify a gang where labels could suffice in theory. We also need to know how many pods comprise a gang (or minimum size of a gang). Encoding that value isn't ideal.
I'd recommend modifying @bsalamat's proposal slightly by including a label selector to identify pods belonging to a gang.

+1 for aiming to have compatibility with existing controllers, Job API at minimum.

Reply:
Used to propose `PodSchedulingGroup` in the [google doc](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit?disco=AAAAB4dD780),
so I totally agree with you to provide such kind of API object in core.

Regarding label selector and 'GangID/GroupID', there're some cases we need to handle:
  - race condition: when a scheduler got a Pod w/o 'Gang/PodSchedulingGroup', how to schedule it? 
    * If schedule the pod and no 'Gang/PodSchedulingGroup' coming, that's OK
    * If schedule the pod, 'Gang/PodSchedulingGroup' coming later and meet 'minAvailable', that's OK
    * If schedule the pod, 'Gang/PodSchedulingGroup' coming later but less than 'minAvailable':
      - if it's because of controller delay, schedule should wait
      - if there's no pod anymore, schedule should handle it by LifeCyclePolicy
    * If does not schedule the pod, how long should we wait?
  - performance: for a batch or run-to-complete workload, the pod come and go in short time; it may take time to
    build relationship between 'Gang/PodSchedulingGroup' and Pod
  - summary/status: if 'Gang/PodSchedulingGroup' support cross Kind/Types, it's better to provide a way
    show the summary/status, such succeeded, running, pending

Here's what I did in kube-arbitrator:
  - race condition: as a batch scheduler, kube-arbitrator do scheduling based on 'job'; so it'll not schedule pods
    until get 'Gang/PodSchedulingGroup'
  - performance: using OwnerReference (controller==true); so 'GangID/GroupID' will be better for performance. 'label selector' is
    kind of short term solution to avoid touching more Kind/Type
  - summary/status: it delegate 'Gang/PodSchedulingGroup' owner to handle summary/status

Another minor comment is the name: as you known, Gang-scheduling is one of requirements, we may enhance this object for other requirements later; so a more
general name maybe better, also including GangLifeCyclePolicy, e.g. changing to PodGroupLifeCyclePolicy. And suppose LifeCyclePolicy is going to be handled
by kube-arbitrator.

Reply:
+1 to general point raised by @bsalamat here!

For naming, how about something simple like SchedulingGroup? Then the gang-ness can be specified in the policy.